backend-rust/src/indexer.rs

```diff
@@ -672,59 +691,86 @@ impl PreparedBlock {
+                finalizers_slot_time.push(block.slot_time);
+                last_finalized.push(block.block_last_finalized.clone());
+
+                new_context.last_finalized_hash = block.block_last_finalized.clone();
+            }
+        }
+
         sqlx::query!(
-            r#"INSERT INTO blocks (height, hash, slot_time, block_time, baker_id, total_amount, total_staked)
-VALUES ($1, $2, $3,
-  (SELECT EXTRACT("MILLISECONDS" FROM $3 - b.slot_time) FROM blocks b WHERE b.height=($1 - 1::bigint)),
-  $4, $5, $6);"#,
-            self.height,
-            self.hash,
-            self.slot_time,
-            self.baker_id,
-            self.total_amount,
-            self.total_staked
+            r#"INSERT INTO blocks
+  (height, hash, slot_time, block_time, baker_id, total_amount, total_staked)
+SELECT * FROM UNNEST(
+  $1::BIGINT[],
+  $2::text[],
+  $3::timestamp[],
+  $4::bigint[],
+  $5::bigint[],
+  $6::bigint[],
+  $7::bigint[]
+);"#,
+            &height,
+            &hash,
+            &slot_time,
+            &block_time,
+            &baker_id as &[Option<i64>],
+            &total_amount,
+            &total_staked
         )
         .execute(tx.as_mut())
-            .await?;
-
-        // Check if this block knows of a new finalized block.
-        // If so, mark the blocks since last time as finalized by this block.
-        let new_context = if self.block_last_finalized != context.last_finalized_hash {
-            let last_height = context.last_finalized_height;
+        .await?;
 
-            let rec = sqlx::query!(
-                r#"
-WITH finalizer
-   AS (SELECT height FROM blocks WHERE hash = $1)
+        let rec = sqlx::query!(
+            r#"
 UPDATE blocks b
-   SET finalization_time = EXTRACT("MILLISECONDS" FROM $3 - b.slot_time),
+   SET finalization_time = EXTRACT("MILLISECONDS" FROM finalizer.slot_time - b.slot_time),
        finalized_by = finalizer.height
-FROM finalizer
-WHERE $2 <= b.height AND b.height < finalizer.height
-RETURNING finalizer.height"#,
-                self.block_last_finalized,
-                last_height,
-                self.slot_time
-            )
-            .fetch_one(tx.as_mut())
-            .await
-            .context("Failed updating finalization_time")?;
-            let new_context = BlockProcessingContext {
-                last_finalized_hash:   self.block_last_finalized.clone(),
-                last_finalized_height: rec.height,
-            };
-            Some(new_context)
-        } else {
-            None
-        };
-        for item in self.prepared_block_items.iter() {
-            item.save(tx).await?;
-        }
+FROM UNNEST($1::BIGINT[], $2::text[], $3::timestamp[]) AS finalizer(height, finalized, slot_time)
+JOIN blocks l ON finalizer.finalized = l.hash
+WHERE b.finalization_time IS NULL AND b.height <= l.height
+"#,
+            &finalizers,
+            &last_finalized,
+            &finalizers_slot_time
+        )
+        .execute(tx.as_mut())
```

Comment:
Can we add a comment on what is happening here please?

Reply:
I tried explaining it, but it is difficult without a drawing board and my ascii art skills are not what they used to be :sweat_smile: